core/google/api/core/retry.py

```diff
@@ -177,6 +180,8 @@ def retry_target(target, predicate, sleep_generator, deadline):
             if not predicate(exc):
                 raise
             last_exc = exc
+            if on_error is not None:
+                on_error()
```

Comment:
Should we pass in the `exc` to `on_error`?

Reply:
Good question:  without arguments, it is a "before retry" hook, rather than an "on error" hook.  I don't know if there would be something in the exception which would influence how it executed (e.g., trailing metadata?).

The motivating usecase (#3775) doesn't care about the exception, which is why I wrote it that way.

Reply:
Let's pass the exception unless you feel strongly we shouldn't?

Reply:
ddcd917